### PR TITLE
libcephfs: Nonblocking io

### PR DESCRIPTION
**Pull request by ffilz**
This replaces https://github.com/ceph/ceph/pull/44991 (async I/O).

This tests as working with Ganesha. I renamed to nonblocking I/O because there was some confusion with Read_Sync_Async that looks a lot better as Read_Sync_Nonblocking. A unit test is included.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ljflores** [comment]
@ffilz looks like this needs a rebase

File: src/osdc/ObjectCacher.h
Comment (adamemerson) [inline src/osdc/ObjectCacher.h:570]: Minor nit, but I'd rather not differentiate things solely by case like this. Maybe name the variable `cfg_block_writes_upfront_`?

File: src/include/buffer.h
Comment (adamemerson) [inline src/include/buffer.h:1290]: In C++, in most cases, `inline` is sufficient in header files and `static inline` can generate spurious warnings.

File: src/include/buffer.h
Comment (adamemerson) [inline src/include/buffer.h:1304]: Instead of copying out of the bufferlist contents into the iovec buffers, could we use a provided iovec to make a bufferlist of static buffers pointing at the iovec buffers and supplying that bufferlist as the target of a read?

File: src/client/Client.cc
Comment (adamemerson) [inline src/client/Client.cc:10603]: Very minor nit, but if you could shorten the headline by one word to get it under 50 characters, it'd work better with a lot of tooling.

File: src/client/Client.cc
Comment (adamemerson) [inline src/client/Client.cc:10891]: You may want to use ceph::coarse_mono_clock or ceph::mono_clock

File: src/client/Client.h
Comment (adamemerson) [inline src/client/Client.h:1394]: Do we have people calling finish directly instead of complete? (And can we be sure nobody calls complete to avoid a double-free?)

File: src/client/Client.h
Comment (adamemerson) [inline src/client/Client.h:1455]: Maybe make this an enum so the logic is easier to folow.

File: src/osdc/ObjectCacher.h
Comment (ffilz) [reply src/osdc/ObjectCacher.h:570]: Sure

File: src/include/buffer.h
Comment (ffilz) [reply src/include/buffer.h:1290]: OK, I'll remove static

File: src/include/buffer.h
Comment (ffilz) [reply src/include/buffer.h:1304]: That sounds like it would be better. This is definitely an area where I was bumping into my lack of understanding so any suggestions on how to change it would be helpful.

File: src/client/Client.cc
Comment (ffilz) [reply src/client/Client.cc:10603]: Headline? I'm confused.

File: src/client/Client.h
Comment (ffilz) [reply src/client/Client.h:1394]: There are some mixups between calling complete and calling finish... I'm not absolutely sure I've got all those right.

File: src/client/Client.h
Comment (ffilz) [reply src/client/Client.h:1455]: Sure

File: src/client/Client.cc
Comment (adamemerson) [reply src/client/Client.cc:10603]: I just meant the first line of the commit. (The title, etc.)

File: src/include/buffer.h
Comment (adamemerson) [reply src/include/buffer.h:1304]: Look at `buffer::create_static` It will make a `buffer::ptr` referencing an already-allocated region of memory.

Then you can shove it into a buffer list with with `push_back`. In principle you should be able to loop over the iovecs and make a bufferlist that will read/write from the same memory.

File: src/client/Client.cc
Comment (adamemerson) [reply src/client/Client.cc:10891]: Make sense.

**Comment by vshankar** [comment]
@ffilz We will target this post reef branching and give it enough time to bake in main branch.

**Comment by mattbenjamin** [comment]
@vshankar I'm not certain, but it looks as if this PR hasn't gotten a single review comment from the cephfs team, in 6 months.  We are depending on starting a process that allows ganesha devs to get ongoing attention from cephfs devs.  Please try to

**Comment by mattbenjamin** [comment]
@ffilz this should probably get a rebase.  More importantly, there are commits which lack your signed-off-by.  Please address that, and try to address the make check failure, after rebasing.

**Comment by ffilz** [comment]
> @ffilz this should probably get a rebase. More importantly, there are commits which lack your signed-off-by. Please address that, and try to address the make check failure, after rebasing.

I did a rebase last week. The unsigned commit is something that hopefully isn't actually needed but I needed it to compile. It's got nothing to do with my code... I will remove that commit before expecting the pull request to be merged.

**Comment by ffilz** [comment]
Clicking on the details for the make check failures doesn't lead to any insight on them...

File: src/include/buffer.h
Comment (ffilz) [reply src/include/buffer.h:1304]: Delving into the code, eventually the bufferlist is utilized by ObjectCacher::_readx which clears (and disposes) of the buffers already in the bufferlist and attaches new buffers. I think what might be going on is a result of data caching. If ceph is maintaining a data cache, and Ganesha wants to get the data into an iovec, a copy must happen.

I think the solution to eliminating this copy is for the API to change so Ganesha can receive buffers from the filesystem and return them when the RPC I/O is complete. Of course in this case, the buffer would be refcounted, so it would remain in the cache, and multiple references could be passed to multiple Ganesha READ requests.

Doing so of course will obligate Ganesha to not change the data (so we could not encrypt in place for example).

Given what it looks like is going on here, I propose we leave this code alone for now, and have a second pass at async/nonblocking I/O where we do work both in Ganesha and cephfs to remove data copying everywhere possible from end to end.

Comment (mattbenjamin) [reply src/include/buffer.h:1304]: just to be clear, we're just agreeing not to attempt copy elision in this PR?

**Comment by ffilz** [comment]
Yes, we will work on removing copies in a separate effort. I think we
actually already have github issues open for eliminating copies.

File: src/include/buffer.h
Comment (ffilz) [reply src/include/buffer.h:1304]: Yes, we will work on removing copies in a separate effort. I think we
actually already have github issues open for eliminating copies.

File: src/client/Client.h
Comment (ffilz) [reply src/client/Client.h:1402]: I will take one more look, but I think I can remove those finish overrides.

**Review by adamemerson — APPROVED** [review]
Looks good to me.

**Comment by vshankar** [comment]
@ffilz I'm back from PTO - this is in my list for reviews.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
Half way through the changes - looking neat.

**Review by vshankar — COMMENTED** [review]
Still going through async fsync and preadv/pwritev changes. Publishing comment for the interim. Looking really good @ffilz.

Will finish up review tomorrow.

**Comment by vshankar** [comment]
@ffilz Were you able to read though my reviews?

This PR is LGTM and I have plans to run the changes through fs suite this week.

**Comment by ffilz** [comment]
> @ffilz Were you able to read though my reviews?
> 
> This PR is LGTM and I have plans to run the changes through fs suite this week.

Thanks for the review.

Yes, I've seen them, sorry was deep in other stuff last week. I will improve the comments, and also move that one hunk.

Is there anything else besides those two things that you need me to do?

Thanks

Frank

**Comment by vshankar** [comment]
> > @ffilz Were you able to read though my reviews?
> > This PR is LGTM and I have plans to run the changes through fs suite this week.
> 
> Thanks for the review.
> 
> Yes, I've seen them, sorry was deep in other stuff last week. I will improve the comments, and also move that one hunk.
> 
> Is there anything else besides those two things that you need me to do?
> 
> Nope. I'll put it through test once those comments are taken care.



> > @ffilz Were you able to read though my reviews?
> > This PR is LGTM and I have plans to run the changes through fs suite this week.
> 
> Thanks for the review.
> 
> Yes, I've seen them, sorry was deep in other stuff last week. I will improve the comments, and also move that one hunk.
> 
> Is there anything else besides those two things that you need me to do?

That's all for now. Will put it through test once you push an update. Thx!

File: src/client/Client.cc
Comment (ffilz) [reply src/client/Client.cc:10785]: True, I had made the previous commit as just adding the new function, and this commit happens to use it. I'm happy to move it if you prefer.

File: src/client/Client.cc
Comment (ffilz) [reply src/client/Client.cc:10714]: OK, I think I've made this a bit more clear, and yea, I think that helps follow the code a bit better.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins retest this plase

**Comment by vshankar** [comment]
> Test results: https://pulpito.ceph.com/vshankar-2023-07-12_07:14:06-fs-wip-vshankar-testing-20230712.041849-testing-default-smithi/
> 
> Will review and update.

@ffilz Nothing stood out when superficially scanning the failed/dead jobs, so, I guess this is almost good to go after I do a thorough analysis of the run. Nice work @ffilz. Thanks!

And, BTW, commit [7ee8bc3](https://github.com/ceph/ceph/pull/48038/commits/7ee8bc392914a0dfb2a15033925a695502f7c335) if new and doesn't have an explanation of why its required. What's going on with that. Did I look over something. @ffilz

**Comment by ffilz** [comment]
> > Test results: https://pulpito.ceph.com/vshankar-2023-07-12_07:14:06-fs-wip-vshankar-testing-20230712.041849-testing-default-smithi/
> > Will review and update.
> 
> @ffilz Nothing stood out when superficially scanning the failed/dead jobs, so, I guess this is almost good to go after I do a thorough analysis of the run. Nice work @ffilz. Thanks!
> 
> And, BTW, commit [7ee8bc3](https://github.com/ceph/ceph/pull/48038/commits/7ee8bc392914a0dfb2a15033925a695502f7c335) is new and doesn't have an explanation of why its required. What's going on with that. Did I look over something. @ffilz

Oh, that last commit was something I needed to get it to compile on my system. I should have avoided pushing that to the repo... I'll remove it and push again.

**Comment by ffilz** [comment]
OK, done that. Sorry.

**Comment by vshankar** [comment]
> OK, done that. Sorry.

Did you forget pushing?

**Comment by ffilz** [comment]
> > OK, done that. Sorry.
> 
> Did you forget pushing?

No, I was in a hurry and didn't pay attention to the error message. It didn't go because the network connection for my VM was bunged up do to the fuse covering my office tripping (and thus killing my switch)...

Should be pushed now...

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
The windows failure is a known issue right now.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#18-July-2023

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
Windows jenkins failure fix: #52427

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
@ffilz please rebase and push - the windows failure has been fixed and merged.

**Pull request merged**
